### ISSUE TITLE
Feedback on auth and identity

### ISSUE THREAD

**Issue by lukepuplett** [dotnet/aspnetcore#11563]
### Is your feature request related to a problem? Please describe.
This is **just feedback**.

I have been having non-stop problems working with authentication, authorization and identity. Some of my woes are written about here.

https://forums.asp.net/p/2156939/6267787.aspx?p=True&t=636970619240804647

### Describe the solution you'd like
I think the design for auth/id is too abstract. I am not asking anyone to rewrite it, only to consider that while middleware and `Add*` and `Use*` extensions are neat, some larger and complex pieces may not suit the pattern.

#### Don't surprise me

These sweet looking extensions often trade instant convenience for important transparency.

At the other end of the scale are APIs designed to force the developer to provide a graph of other well named pieces that, while fiddly on first use, serve to setup an accurate mental model.

Since becoming open-source I feel that we've sometimes forgotten to design instructive APIs.

#### "Kitchen sink" extensibility

I think the framework overreaches with its aim to be extensible so as to take care of everything for the programmer, including linking identities, emailing, ensuring emails are confirmed and 2FA, on top of OAuth flows and reading user information from OAuth APIs.

#### Less is more

A few years back I took the MS code and rewrote the auth OWIN middleware as I was having problems and couldn't follow the MS/OS code. And before that, I wrote my OAuth stuff all in controller actions, which was actually the easiest to maintain since the flow of request/response/backchannel was so easy to read and debug.

I wonder if it would be better to supply components to make it easier for developers to formulate the requests, encode/protect information on the URL, build cookies, read cookies into a principal etc. and then demonstrate putting it all together for a few scenarios.

Such a set of parts might also survive the future a little better.

Thanks.

**Comment by lukepuplett** [comment]
Here are a few example problems that I think are rooted in design choices.

https://stackoverflow.com/q/56743645/107783
https://stackoverflow.com/a/56551438/107783
https://github.com/aspnet/AspNetCore/issues/11146

I'm not looking for specific answers, there are other forums for that. These problems were tough to troubleshoot because they resulted from choices I didn't expressly make in my code.

**Comment by lukepuplett** [comment]
Coincidently, David Fowler posted a Gist showing how few lines of code are needed to create a working microservice on ASP.NET Core 3.0.

https://gist.github.com/davidfowl/1fdbc02ef9511e2729a914a8c9f3ab27#gistcomment-2959243

I pushed back on this as I think it celebrates the wrong metric. If you had somehow had metrics on the number of people getting stuck using auth, that might be more to celebrate.

I've finally got Google login auth working after weeks of battles, ending in a problem with the order of extension methods.

Again, these extension methods and the names of the pieces obscure reality and prevent me from building an accurate mental model.

Perhaps calling `IApplicationBuilder` would be better as `IPipelineBuilder` and designs using extensions with `Then` and would help here, or a design that means the calls have to be nested, like the pipeline its constructing. The term `Use` sounds like I can just chuck it in anywhere.

This may not work, but it shows my point.

    webBuilder.ConfigurePipeline(pipeline =>
            {
                pipeline.StartWith().Routing()
                    .ThenCall().Endpoints(route =>
                    {
                        route.MapGet("/", context => context.Response.WriteAsync("Hello world"));
                    });
            });

Or

    webBuilder.ConfigurePipeline(pipeline =>
            {
                pipeline.UseRouting()
                    .Then().UseEndpoints(route =>
                    {
                        route.MapGet("/", context => context.Response.WriteAsync("Hello world"));
                    });
            });

Other "set pieces of pipeline" could be built in more specialized builders that make it hard to get it all in the wrong order, and then grafted onto the main app pipeline using clear language.

The idea is to expose how its actually working. Make the innards discoverable.

Challenge: just as Apple shipped the worlds most advanced phone in 2007, without a user manual, could an API be built so well that it doesn't need documentation?

A good website doesn't need documentation, since websites only offer up _legal_ next steps. It's amazing when you consider that a website is really just helping the user build the correct POST request to instruct some person to send a pizza to your door.
